src/ServiceControl/Monitoring/HeartbeatMonitoring/HeartbeatTypesArrayToInstanceConverter.cs

```diff
@@ -0,0 +1,107 @@
+    public class HeartbeatTypesArrayToInstanceConverter : JsonConverterFactory
+    {
+        public override bool CanConvert(Type typeToConvert) => typeToConvert == typeof(RegisterEndpointStartup) ||
+                                                               typeToConvert == typeof(EndpointHeartbeat) ||
+                                                               typeToConvert ==
+                                                               typeof(RegisterPotentiallyMissingHeartbeats);
+
+        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options) =>
+            typeToConvert switch
+            {
+                _ when typeToConvert == typeof(RegisterEndpointStartup) => new Converter<RegisterEndpointStartup>(options),
+                _ when typeToConvert == typeof(EndpointHeartbeat) => new Converter<EndpointHeartbeat>(options),
+                _ when typeToConvert == typeof(RegisterPotentiallyMissingHeartbeats) => new Converter<RegisterPotentiallyMissingHeartbeats>(options),
+                _ => throw new NotSupportedException()
+            };
+
+        sealed class Converter<TValue>(JsonSerializerOptions options) : JsonConverter<TValue>
+        {
+            // Currently we want to rely on deserializing the actually value by using the standard json serializer
+            // To make sure we are not recursively invoking the converter we need to remove it from the options
+            // Removing from the options is currently only possible when the converter is added to the options directly
+            // and not supported when the converter is added to the type
+            readonly JsonSerializerOptions optionsWithoutCustomConverter = options.FromWithout<HeartbeatTypesArrayToInstanceConverter>();
+
+            public override TValue? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                var readArray = false;
+                if (reader.TokenType == JsonTokenType.StartArray)
+                {
+                    readArray = true;
+                    reader.Read();
+
+                    if (reader.TokenType != JsonTokenType.StartObject)
+                    {
+                        throw new JsonException();
+                    }
+                }
+
+                var value = JsonSerializer.Deserialize<TValue>(ref reader, optionsWithoutCustomConverter);
+
+                if (!readArray)
+                {
+                    return value;
+                }
+
+                if (reader.TokenType != JsonTokenType.EndObject)
+                {
+                    throw new JsonException();
+                }
+
+                _ = reader.Read();
+
+                if (reader.TokenType != JsonTokenType.EndArray)
+                {
+                    throw new JsonException();
+                }
+
+                _ = reader.Read();
```

Comment:
What is the purpose of this read?

Reply:
It is reading the EndArray which has to be consumed